### PR TITLE
Content(Hai): add optical jammers and put them on recent/military ships

### PR DESCRIPTION
**Pull request by Hurleveur**
**Content (Artwork / Missions / Jobs)**

## Summary
Make use of the recently added "optical jamming" attribute and of hai jammers.
These new jammers could easily be added to all ships but I added them to ships made after the Hai-UHai conflict and military ships.
Not all military ships have had these added to them, because they do consume a bunch of energy, so dont pair well except if you yourself use a bunch of trackers which dont consume much energy to shoot.

<details>
<summary>Hai Cuttlefish Jammer</summary>

image

</details>

<details>
<summary>The Unfettered have a 30% stronger, 50% more energy heavy variant of it.</summary>

image

</details>

(Also fixes the Unfettered large fleet not having sea scorpions anymore, probably due to a mis-merge)
The custom unfettered shield beetle build uses value detectors too.

## Save File
Here in Wah Ki reporting on the situation: how effective are these jammers?
[Emile.Lancer.JD.from.the.void.txt](https://github.com/endless-sky/endless-sky/files/10328611/Emile.Lancer.JD.from.the.void.txt)

### CODE REVIEW

**Comment by Hurleveur** [comment]
I'm really not sure how this is balance wise but please don't compare these too much to the normal jammers because those completely suck (even human merchants tell you that) and this is Hai tech, too. This should only be looked at in terms of performances on the battlefield: if you shoot from 1k range 8 trackers at these ships, how many hit them and how many deviate? That's what interesting to know to decide on it.

**Comment by Hurleveur** [comment]
Just did some tests with a cold beetle (as in no heat) + 4 jammers and it was relatively easy to dodge the trackers, quite a few lost their focus on me but not too much either. I'm happy with this performance.

**Review by Zitchas — COMMENTED** [review]
These look good. It's nice to have some optical jammers in the game, and based on Hurleveur's testing, these sound like they're in a decent spot.

That being said, I did some testing of my own.

[Hai tracker testing.txt](https://github.com/endless-sky/endless-sky/files/10329034/Hai.tracker.testing.txt)

[Rebbia Rampart~3023-07-25 Cuttlefish Testing.txt](https://github.com/endless-sky/endless-sky/files/10329074/Rebbia.Rampart.3023-07-25.Cuttlefish.Testing.txt)

Flying an Osprey with negative idle heat (although some heat does build up with agressive flying and afterburner use) and a single cuttlefish jammer, and I'm not really noticing much. Maybe a *slight* increase in the number of trackers that lose target, and a bit more than waver off course and then regain their lock. But only barely noticeable. I'm not sure it'd be enough for me to want to use one.

Just to be excessive, changed up my ship to be ice cold even when moving (so, idles at -2k), with 14 cuttlefish and enough battery to run it for quite a while. I'm seeing s omething like 10% failure rate in the trackers? Maybe 20%? This is low enough that I'm downloading a fresh exe in case mine accidentally didn't have the optical jammer code or something.

File: data/hai/hai outfits.txt [hunk 1]
```diff
@@ -399,6 +399,27 @@ outfit "Chameleon Anti-Missile"
 
 
 
+outfit "Hai Cuttlefish Jammer"
+	category "Systems"
+	cost 260000
+	thumbnail "outfit/hai jammer"
+	"mass" 10
+	"outfit space" -7
+	"optical jamming" 20
```
[the remark this answers is not in the document]
Comment (Hurleveur) [reply data/hai/hai outfits.txt:408]: human radars suck so much nobody uses them ever I don't think there even is a discussion about the fact they suck yeah
I think they should be massively buffed

also do take into account these jammers take more mass and actively take energy all the time so they ought to be better

**Comment by Hurleveur** [comment]
They don't scale as well as I'd want them to but you should notice way more of an effect than that, except if you're at point blank range.
10-20% of them going away is default behavior

**Comment by Zitchas** [comment]
Actually, I think the problem is the IR tracking. I commented that out of the Hai Trackers, and suddenly my 15 cuttlefish went from jamming <20% to jamming >90%. So yeah, they work. I pity anything else in the game that uses optical tracking.

Without the IR tracking, just a single cuttlefish had something like a 30-40% jamming rate. Which, I'll be honest, is actually rather nice to see. Jammers *should* have some use. In this context, I'd say these are a little on the powerful side. Not massively, so, just a bit. I haven't tested values, but something like, say, 13 & 20 instead of 20 & 27.

That being said, that's an important thing to note: While these are a little effective against Hai Trackers that they are designed for, this also has the effect of rendering anything else that uses optical tracking (which has generally been the tracking mechanism of choice for things that have extremely good tracking but tend to be slow, like torpedoes) extremely vulnerable. 

This isn't necessarily a problem, just something that we need to pay attention. The effects of this on overal gameplay extends to more than just Hai Trackers. In fact, Hai Trackers are, probably, the optically guided thing they are *least* useful for with the current mechanics.

**Comment by Hurleveur** [comment]
Looking at code the infrared tracking's chance of being set off does not depend on heat... bruh?
Only the chance of it losing a lock does. Also with the big range of the trackers they often are getting x3 to their tracking so they have less chances of losing a lock.

**Comment by Zitchas** [comment]
Yeah, I'm looking at the heat tracking code too. In terms of my ship, my loadout was at -2k heat idle so that even under full propulsion load it should never actually go above 0. So really, solidly, firmly lock it to 0  heat. I wouldn't want to actually run a ship like this - the cooling takes up so much extra space... But it was an attempt to nullify the IR tracking aspect of the trackers. As demonstrated by my testing, though, there is vastly more than a 5% difference...

So where does temperature come into play, then? With such a cold ship, shouldn't it be hitting the 5% lower bound for IR tracking?

**Comment by Zitchas** [comment]
So the IR homing of .4 overrides the "target has a heat of 0, so tracking should be a 5% chance" part?

**Comment by Hurleveur** [comment]
Temperature only matters to determine if the trackers keep a lock on you, with optical.
If they lost both optical and heat lock on you then they have a chance to move away, if they fail yet another test; dependent on their tracking quality. When combining 2 tracking types, they both have to fail for the missile to move away from pointing your ship.
Since the infrared tracking is .4 it has 60% chance of failing.
The optical is 80%, and is then affected by your size and jamming.
This means that in the best case scenario, on average, you will only have the missile move away 50% of the time and heading to you the other 50%. If you dont move fast enough it'll get you because of how good it is at homing (level 4)

**Comment by Zitchas** [comment]
I'm not seeing much of that. My Osprey is relatively fast, and generally everything is following it and responding to its course changes.

**Comment by Hurleveur** [comment]
You wouldn't notice it if the tracker was not moving your way for a frame, then heading straight and then heading towards you, that's the issue here I think.
You only notice it when they keep on moving away.

**Comment by Zitchas** [comment]
[Hai tracker testing.txt](https://github.com/endless-sky/endless-sky/files/10329252/Hai.tracker.testing.txt)

For those interested, this updated test file (just drop it in your data folder before running the game) has two test missions now. One is a tracker mission that creates a stationary SB with a single tracker pod. The second tester mission has the same stationary SB, except with a full loadout of 8 nuclear missiles. These have an optical tracking value of 1, and a radar tracking of .5 So they make an effect alternate thing to test this jammer against.


Case in point, I find it odd that my ship - which has no radar jammers at all, so just the system's distraction of the asteroids - The nuclear missiles with 1 optical and .5 radar fare poorer against my Osprey with 1 cuttlefish and 0 heat than the tracker with it's .8 optical and .4 IR does. 

This really doesn't seem right.

**Comment by Hurleveur** [comment]
Something is going wrong with that indeed...
Let's test a little thing, I made the moving away twice as strong. I think a PR changing how missiles work will be needed before this one can be merged.

**Comment by Zitchas** [comment]
Given the... problems... with how missiles work; I'd suggest just tweaking the values down a bit and merge anyway. We want these optical jammers, and they fit the lore quite well, so that's not in question. And it gives us stuff to use in testing for the rework of how missiles work.

Worst case scenario they join the ranks of the radar jammer in things that do a little something but nearly as much as people want. At least it is in-game where it should be and not just a draft PR hanging around waiting for nebulous future stuff.

File: data/hai/hai outfits.txt (hunk 1 above)
Comment (Zitchas) [inline data/hai/hai outfits.txt]:
```suggestion
	"optical jamming" 14
```

File: data/hai/hai outfits.txt [hunk 2]
```diff
@@ -399,6 +399,27 @@ outfit "Chameleon Anti-Missile"
...
+	"energy consumption" .8
+	description "When the Hai first started fighting each others there quickly was a need for a counter to trackers. The Cuttlefish Jammer will notably disturb the tracker's lock to the ship, by presenting it with a seemingly better target. This jammer is mostly needed for larger ships, that often cannot field enough anti-missiles to effectively counter trackers."
+
+outfit "Hai Octopus Jammer"
+	category "Systems"
+	cost 350000
+	thumbnail "outfit/hai jammer overclocked"
+	"mass" 10
+	"outfit space" -7
+	"optical jamming" 26
```
Comment (Zitchas) [inline data/hai/hai outfits.txt]:
```suggestion
	"optical jamming" 21
```

**Review by Zitchas — APPROVED** [review]
Just to make it formal, I'd like to suggest these values for the optical jammers. They're a bit lower so they don't overwhelm every other optically guided thing in the game (including nuclear missiles), but seems to be enough to have a noticeable and beneficial effect. These are suggestions, not requirements, and as such I'm approving this as is.

Given that testing has revealed some problems with how jamming is handled, and how IR tracking in particular works, I'd like to suggest we consider merging this so that we have optical jammers in game that we can use to test any future reworks (not to mention dispense with having a PR of desireable content hanging over our heads taking up space). 

In terms of lore, these make sense, and fit into the existing material we have in game. I'd like a little more description, but that isn't necessary either.

edit: marked my two suggestions as resolved as you implemented slight nerfs while I was posting this.

**Review by MasterOfGrey — APPROVED** [review]

File: data/hai/hai outfits.txt [hunk 3]
```diff
@@ -399,6 +399,27 @@ outfit "Chameleon Anti-Missile"
...
 
+outfit "Hai Cuttlefish Jammer"
+	category "Systems"
+	cost 260000
+	thumbnail "outfit/hai jammer"
+	"mass" 10
+	"outfit space" -7
+	"optical jamming" 15
+	"energy consumption" .8
+	description "When the Hai first started fighting each others there quickly was a need for a counter to trackers. The Cuttlefish Jammer will notably disturb the tracker's lock to the ship, by presenting it with a seemingly better target. This jammer is mostly needed for larger ships, that often cannot field enough anti-missiles to effectively counter trackers."
```
Comment (EjoThims) [inline data/hai/hai outfits.txt:410]:
```suggestion
	description "When the Hai first started fighting each other, the need for a counter to trackers quickly became apparent. The Cuttlefish Jammer can noticebly disturb a tracker's lock to the ship by presenting it with a seemingly better, albeit illusory, target. This jammer is most effective on larger ships that cannot field enough anti-missiles to effectively counter them from all directions."
```
I feel like we could have a better opener here if it were more specific. @MasterOfGrey do we have the initiation of ongoing conflict with the Unfettered mapped out? At least in terms of when/who fighting started?

File: data/hai/hai outfits.txt [hunk 4]
```diff
@@ -399,6 +399,27 @@ outfit "Chameleon Anti-Missile"
...
+
+outfit "Hai Octopus Jammer"
+	category "Systems"
+	cost 350000
+	thumbnail "outfit/hai jammer overclocked"
+	"mass" 10
+	"outfit space" -7
+	"optical jamming" 20
+	"energy consumption" 1.2
+	description "The Octopus Jammer is a more aggressive version of the Cuttlefish Jammer, intended for short raidings by the Unfettered. It is able to jam trackers more but at a much higher energy cost, meaning it makes them even more vulnerable to ion weaponry than usual jammers do."
```
Comment (EjoThims) [inline data/hai/hai outfits.txt:420]:
```suggestion
	description "The Octopus Jammer is a more aggressive adaptation of the Cuttlefish Jammer, intended for short raids by the Unfettered. With several safety circuits bypassed, it is able to more effectively deceive incoming trackers more, but the much higher associated energy cost makes any ship carrying this even more vulnerable to ion weaponry."
```

**Review by EjoThims — CHANGES_REQUESTED** [review]

**Comment by Hurleveur** [comment]
That would make you more vulnerable to jammer because of the infrared lock

**Review by EjoThims — CHANGES_REQUESTED** [review]

File: data/hai/hai outfits.txt [hunk 5]
```diff
@@ -399,6 +399,27 @@ outfit "Chameleon Anti-Missile"
...
 
+outfit "Hai Cuttlefish Jammer"
+	category "Systems"
+	cost 260000
+	thumbnail "outfit/hai jammer"
+	"mass" 10
+	"outfit space" -7
+	"optical jamming" 15
+	"energy consumption" .8
+	description "When the Hai first started fighting each other, the need for a counter to trackers quickly became apparent. The Cuttlefish Jammer can noticebly disturb a tracker's lock to the ship by presenting it with a seemingly better, albeit illusory, target. This jammer is most effective on larger ships that cannot field enough anti-missiles to effectively counter them from all directions."
```
Comment (Amazinite) [inline data/hai/hai outfits.txt:410]:
```suggestion
	description "When the Hai first started fighting each other, the need for a counter to Trackers quickly became apparent. The Cuttlefish Jammer can noticeably disturb a Tracker's lock to the ship by presenting it with a seemingly better, albeit illusory, target. This jammer is most effective on larger ships that cannot field enough anti-missiles to effectively counter missiles from all directions."
```

File: data/hai/hai outfits.txt [hunk 6]
```diff
@@ -399,6 +399,27 @@ outfit "Chameleon Anti-Missile"
...
+
+outfit "Hai Octopus Jammer"
+	category "Systems"
+	cost 350000
+	thumbnail "outfit/hai jammer overclocked"
+	"mass" 10
+	"outfit space" -7
+	"optical jamming" 20
+	"energy consumption" 1.2
+	description "The Octopus Jammer is a more aggressive adaptation of the Cuttlefish Jammer, intended for short raids by the Unfettered. With several safety circuits bypassed, it is able to more effectively deceive incoming trackers, but the much higher associated energy cost makes any ship carrying this even more vulnerable to ion weaponry."
```
Comment (Amazinite) [inline data/hai/hai outfits.txt:420]:
```suggestion
	description "The Octopus Jammer is a more aggressive adaptation of the Cuttlefish Jammer, intended for short raids by the Unfettered. With several safety circuits bypassed, it is able to more effectively deceive incoming Trackers, but the much higher associated energy cost makes any ship carrying this even more vulnerable to ion weaponry."
```

**Pull request merged**